aggregators/aggregator_test.go

```diff
@@ -1075,163 +1075,444 @@ func TestHarvest(t *testing.T) {
 
 func TestAggregateAndHarvest(t *testing.T) {
 	txnDuration := 100 * time.Millisecond
-	batch := modelpb.Batch{
-		{
-			Event: &modelpb.Event{
-				Outcome:  "success",
-				Duration: uint64(txnDuration),
-			},
-			Transaction: &modelpb.Transaction{
-				Name:                "foo",
-				Type:                "txtype",
-				RepresentativeCount: 1,
-			},
-			Service: &modelpb.Service{Name: "svc"},
-			Labels: modelpb.Labels{
-				"department_name": &modelpb.LabelValue{Global: true, Value: "apm"},
-				"organization":    &modelpb.LabelValue{Global: true, Value: "observability"},
-				"company":         &modelpb.LabelValue{Global: true, Value: "elastic"},
-				"mylabel":         &modelpb.LabelValue{Global: false, Value: "myvalue"},
-			},
-			NumericLabels: modelpb.NumericLabels{
-				"user_id":        &modelpb.NumericLabelValue{Global: true, Value: 100},
-				"cost_center":    &modelpb.NumericLabelValue{Global: true, Value: 10},
-				"mynumericlabel": &modelpb.NumericLabelValue{Global: false, Value: 1},
-			},
-		},
+
+	runTest := func(t *testing.T, batch modelpb.Batch, expected []*modelpb.APMEvent) {
+		var events []*modelpb.APMEvent
+		agg, err := New(
+			WithDataDir(t.TempDir()),
+			WithLimits(Limits{
+				MaxSpanGroups:                         1000,
+				MaxSpanGroupsPerService:               100,
+				MaxTransactionGroups:                  100,
+				MaxTransactionGroupsPerService:        10,
+				MaxServiceTransactionGroups:           100,
+				MaxServiceTransactionGroupsPerService: 10,
+				MaxServices:                           10,
+			}),
+			WithProcessor(sliceProcessor(&events)),
+			WithAggregationIntervals([]time.Duration{time.Second}),
+		)
+		require.NoError(t, err)
+		require.NoError(t, agg.AggregateBatch(
+			context.Background(),
+			EncodeToCombinedMetricsKeyID(t, "ab01"),
+			&batch,
+		))
+		require.NoError(t, agg.Close(context.Background()))
+
+		assert.Empty(t, cmp.Diff(
+			expected,
+			events,
+			cmpopts.IgnoreTypes(netip.Addr{}),
+			cmpopts.SortSlices(func(a, b *modelpb.APMEvent) bool {
```

Comment:
I see that a lot of effort is put into writing a good sort func here, but we could cut some corners here as perf isn't critical. I was thinking something along the lines of serializing it and comparing the output, and I found that we already do similar stuff in apm-data: https://github.com/elastic/apm-data/blob/81d77648a3b17d4e52859110e233e40683fcdfe3/input/otlp/metrics_test.go#L1146-L1148

This is proto and perf should be better than json.

Reply:
Thanks for the suggestion, I don't think we are much interested in performances here as we sort such a small set. Updated in [396c6e5](https://github.com/elastic/apm-aggregation/pull/191/commits/396c6e5e6f19be347fec418b84d59531ee7e352e)